virttest/utils_disk.py

```diff
@@ -600,6 +601,96 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+    :return: Tuple (fstype, mountpoint)
+    """
+    mount_list = session.cmd_output('cat /proc/mounts')
+    pattern = r'%s\s(.+?)\s(.+?)\s' % partition
+    mount_info = re.search(pattern, mount_list)
+    if mount_info:
+        mountpoint = mount_info.group(1)
+        fstype = mount_info.group(2)
+        return (fstype, mountpoint)
+    else:
+        logging.error('partition should not be mounted, mounted'
+                      'list as:\n%s' % mount_list)
+
+
+def resize_filesystem_linux(session, partition, size):
+    """
+    resize file system in linux guest.
+    support enlarge and shrink.
+
+    :param session: session object to guest.
+    :param partition: disk partition, like /dev/sdb1.
+    :param size: resize file system to size, unit is M.
+                 support transfer size='MAX_AVAILABLE_SIZE',
+                 enlarge to maximun available size.
+    """
+    fstype, mountpoint = get_fstype_mpoint_linux(session, partition)
+    if fstype == 'xfs':
+        if size == 'MAX_AVAILABLE_SIZE':
+            resize_fs_cmd = 'xfs_growfs -d %s' % mountpoint
+        else:
+            resize_fs_cms = 'xfs_growfs -D %s %s' % (size, mountpoint)
+    else:
+        if is_mount(partition, dst=mountpoint, fstype=fstype, session=session):
+            umount(partition, mountpoint, fstype=fstype, session=session)
+
+        session.cmd('e2fsck -f %s' % partition)
+
+        if size == 'MAX_AVAILABLE_SIZE':
+            resize_fs_cmd = 'resize2fs %s' % partition
+        else:
+            size = utils_misc.normalize_data_size(size, order_magnitude="M")
+            size = int(size.split(".")[0]) - 1
+            resize_fs_cmd = 'resize2fs %s %sM' % (partition, size)
+    session.cmd(resize_fs_cmd)
+
+    if not is_mount(partition, dst=mountpoint, fstype=fstype, session=session):
+        mount(partition, mountpoint, fstype=fstype, session=session)
+
+
+def resize_partition_linux(session, partition, size, start='0M'):
+    """
+    resize partition in linux guest.
+
+    :param session: session object to guest.
+    :param partition: partition that to be shrunk, like /dev/sdb1.
+    :param size: resize partition to size, unit is M.
+    :param start: mkpart primary start place, default is 0M.
+    """
+    fstype, mountpoint = get_fstype_mpoint_linux(session, partition)
+    if is_mount(partition, dst=mountpoint, fstype=fstype, session=session):
+        umount(partition, mountpoint, fstype=fstype, session=session)
+
+    part_info = re.match(r'(/dev/.*)(\d+)', partition)
+    dev_name = part_info.group(1)
+    part_num = part_info.group(2)
+
+    parted_cmd = 'parted -s %s print' % dev_name
+    rm_part_cmd = ' '.join((parted_cmd, 'rm %s' % part_num))
+    session.cmd(rm_part_cmd)
+
+    size = float(utils_misc.normalize_data_size(size, order_magnitude="M",
+                                                factor="1000"))
+    resizepart_cmd = ' '.join((parted_cmd, 'mkpart primary {0} {1}M'))
```

Comment:
How to confirm the part type is "primary"?

Reply:
updated the code, get the partition type first~
thanks~